### ISSUE TITLE
First element of array returned by explode() is never null

### ISSUE THREAD

**Issue by sanmai** [vimeo/psalm#7101]
https://psalm.dev/r/154ffd42b2
```php
/** @var string $q */
$a = explode(',', $q);

return $a[0] ?? '';
```

Compare with:

https://3v4l.org/pgbW6

**Comment by weirdan** [comment]
What do you expect though? Psalm already infers the result as `non-empty-list<string>`: https://psalm.dev/r/2b84c861bb

**Comment by AndrolGenhald** [comment]
@weirdan It looks like a more general issue where the `??` operator is redundant, but this isn't detected by psalm: https://psalm.dev/r/67ae043a26

**Comment by sanmai** [comment]
@weirdan You're correct, that's `non-empty list<string>`. That's `TypeDoesNotContainNull` is that not?

**Comment by weirdan** [comment]
Or `UnevaluatedCode`.

**Comment by orklah** [comment]
I added the help wanted tag, that should be quite easy to do.

Coalesce operator is performed in CoalesceAnalyzer. It should be easy to retrieve the type of the first operand and check if the union contains null.

The only caveat will be the case where nullsafe were used. For example

`$a?->alwaysReturnNull()->anotherMethod() ?? ''` is likely to be resolved by psalm as not containing null because the way Psalm handles nullsafe is really not good: https://github.com/vimeo/psalm/issues?q=is%3Aissue+is%3Aopen+label%3Anullsafe
But that's another issue completely.

**Comment by kkmuffme** [comment]
This isn't an issue with coalesce, but we have the exact same issue when using the classic variant with isset:
https://psalm.dev/r/73de65926d
